proposed/caller-unsafe.md

```diff
@@ -0,0 +1,60 @@
+
+## Goals
+
+The existing unsafe system does not clearly identify which methods need hand-verification to use safely, and it doesn't indicate which methods claim to provide that verification (and produce a safe API).
+
+We want to achieve all of the following goals:
+
+* Clearly annotate which methods require extra verification to use safely
+* Clearly identify in source code which methods are responsible for performing extra verification
+* Provide an easy way to identify in source code if a project doesn't use unsafe code
+
+## Proposal
+
+We need to be able to annotate code as unsafe, even if it doesn't use pointers.
+
+Mechanically, this would be done with a new attribute:
+
+```C#
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Constructor)]
+public sealed class RequiresUnsafe : System.Attribute
+{
+    public WarnByDefault { get; init; } = true;
+}
+```
+
+This would be the dual of the existing `unsafe` context. The `unsafe` context eliminates unsafe errors, indicating that the code outside of the unsafe context should be considered safe. The `RequiresUnsafe` attribute does the opposite: it indicates that the code inside is unsafe unless specific obligations are met. These obligations cannot be represented in C#, so they must be listed in documentation and manually verified by the user of the unsafe code. Only when all obligations are discharged can the code be wrapped in an `unsafe` context to eliminate any warnings.
+
+The `WarnByDefault` flag is needed for backwards-compatibility. If an existing API is unsafe, adding warnings would be a breaking change. If `WarnByDefault` is set to `false`, warnings are not produced unless a project-level property, `ShowAllRequiresUnsafeWarnings`, is set to true.
+
+### Implementation
+
+Since only warnings are an output of the above design, the feature could be implemented as an analyzer for C#.
+
+### Definition of Unsafe
+
+`RequiresUnsafe` is only useful if there is an accepted definition of what is considered unsafe. For .NET there are two properties that we already consider safe code to preserve:
+
+* Memory safety
+* No access to uninitialized memory
+
+In this document **memory safety** is strictly defined as: safe code can never acquire a reference to memory that is not managed by the application. "Managed" here does not refer to solely to heap-allocated, garbage collected memory, but also includes stack-allocated variables that are considered allocated by the runtime.
+
+No access to uninitialized memory means that all managed memory is either never read before it has been initialized by C# code, or it has been initialized to a zero value.
+
+Some examples of APIs or features that are unsafe due to exposing uninitialized memory include:
```

Comment:
I think `StructLayout.Explicit` can be unsafe as well.

Reply:
What makes Explicit structs memory unsafe? Presumably, the runtime already guarantees that you can't mess up with the GC references (wrong alignment or aliased with some other type)

Reply:
Do we count accessing improperly aligned memory as unsafe? It is reasonable for APIs to expect their `ref T` to be correctly aligned if they need that to work & not check it. If so (which I think would be reasonable), then we should treat it as `unsafe`, and also any use of any such field that isn't guaranteed to be correctly aligned. Explicit layout also allows accessing padding values as non-padding, which could read uninitialized memory.

Reply:
> Explicit layout also allows accessing padding values as non-padding, which could read uninitialized memory.

not sure I follow, what uninitialized memory with explicit layout you're referring to?

Reply:
> not sure I follow, what uninitialized memory with explicit layout you're referring to?

Hmm, I think you're right about this actually. I wasn't thinking clearly enough about definite assignment rules with this (I think I thought that you could just write 1 field in a union-like one & then immediately read out another, but you obviously can't do that). But the other concerns I mentioned still remain valid I think.

Reply:
> What makes Explicit structs memory unsafe? Presumably, the runtime already guarantees that you can't mess up with the GC references (wrong alignment or aliased with some other type)

You're right, aliasing objects and value types doesn't seem to work anymore:

>  System.TypeLoadException: Could not load type 'S' from assembly 'console, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' because it contains an object field at offset 0 that is incorrectly aligned or overlapped by a non-object field.